### DISCUSSION TITLE
Assigning a task to a particular core of esp32

### DISCUSSION THREAD

**Discussion by sergio-caracas** [arkhipenko/TaskScheduler#166] [Q&A]
Hello everyone, 

I am trying to get two taskscheduler tasks to run in parallel, one in each of the two cores of the esp32, is it possible?

I have two audio filters (left and right channel) but the algorithm is still very heavy and the sample rate I am achieving is quite low.

Thanks in advance if you could share any suggestions!

Cheers,
Sergio

**Comment by arkhipenko** [comment]
The only way what you are trying to do is possible is if you run separate schedulers per channel in a dedicated freertos tasks pinned to cores.

As a cooperative scheduler, TaskScheduler does not supper true "parallel" execution as all tasks are running in the same thread.

Conceptually this setup is described here:
https://github.com/arkhipenko/dnd



Sent from my mobile implant.
Apologies for autocorrect.
________________________________
From: sergio-caracas ***@***.***>
Sent: Saturday, May 13, 2023 4:07:14 AM
To: arkhipenko/TaskScheduler ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [arkhipenko/TaskScheduler] Assigning a task to a particular core of esp32 (Discussion #166)


Hello everyone,

I am trying to get two taskscheduler tasks to run in parallel, one in each of the two cores of the esp32, is it possible?

I have two audio filters (left and right channel) but the algorithm is still very heavy and the sample rate I am achieving is quite low.

Thanks in advance if you could share any suggestions!

Cheers,
Sergio

**Reply by sergio-caracas** [reply]
Thank you very much Mr Arkhipenko! 

Very interesting and many things became clearer for me after reading your message and through all the code in your dnd project!

I need to get about 50kHz filter cycle execution on both channels, but my filters are crashing at what I guess is about 200 microseconds (5 kHz).

I shall check if RTOS task granularity can have ticks shorter than 1 millisecond. 

Will keep looking for more information.

Cheers and many thanks,
Sergio

**Comment by teuthid** [comment]
But you can execute any function on a given CPU with esp_ipc_call().
https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/system/ipc.html#_CPPv412esp_ipc_call8uint32_t14esp_ipc_func_tPv

**Comment by arkhipenko** [comment]
There is no inter-process by design.
Everything runs in the same thread.
You are thinking about this in a multi-threading, multi-process way, and it is not that.

**Comment by teuthid** [comment]
You are completely wrong. Please read the documentation I linked to. Thank
you for your co-operation.

**Reply by arkhipenko** [reply]
@teuthid - you are welcome to submit a working prototype PR using the IPC. 
Once that is done and working, and not breaking TS conceptually, then you can use pronouncements like "completely wrong".

**Comment by teuthid** [comment]
@arkhipenko I've used esp_ipc_call() many times in TaskCallback functions to evenly distribute the load across the two cpus. This does not require any modifications to the TaskScheduler. I don't really understand why you consider this to be such a big challenge. Knowledge of ESP-IDF is enough.

**Comment by arkhipenko** [comment]
You can try and enable microsecond precision for TaskScheduler. But you might just hit the hardware limit actually.

Sent from my mobile implant.
Apologies for autocorrect.
________________________________
